ext/json/ext/generator/generator.c

```diff
@@ -25,96 +25,77 @@ static ID i_to_s, i_to_json, i_new, i_pack, i_unpack, i_create_id, i_extend;
-        else {
-            rb_raise(rb_path2class("JSON::GeneratorError"), "source sequence is illegal/malformed utf-8");
-        }
-
-        for (i = 1; i < ch_len; i++) {
-            ch = (ch<<6) | (in_utf8_str[pos+i] & 0x3F);
-        }
-
-        /* JSON policy */
-        should_escape =
-            (ch < 0x20) ||
-            (ch == '"') ||
-            (ch == '\\') ||
-            (out_script_safe && (ch == '/')) ||
-            (out_script_safe && (ch == 0x2028)) ||
-            (out_script_safe && (ch == 0x2029));
-
+    for (pos = 0; pos < len;) {
+        unsigned char ch = ptr[pos];
         /* JSON encoding */
-        if (should_escape) {
-            if (pos > beg) {
-                fbuffer_append(out_buffer, &in_utf8_str[beg], pos - beg);
-            }
-
-            beg = pos + ch_len;
+        if (escape_table[ch]) {
+#define FLUSH_POS(bytes) if (pos > beg) { fbuffer_append(out_buffer, &ptr[beg], pos - beg); } pos += bytes; beg = pos;
             switch (ch) {
-                case '"':  fbuffer_append(out_buffer, "\\\"", 2); break;
-                case '\\': fbuffer_append(out_buffer, "\\\\", 2); break;
-                case '/':  fbuffer_append(out_buffer, "\\/", 2); break;
-                case '\b': fbuffer_append(out_buffer, "\\b", 2); break;
-                case '\f': fbuffer_append(out_buffer, "\\f", 2); break;
-                case '\n': fbuffer_append(out_buffer, "\\n", 2); break;
-                case '\r': fbuffer_append(out_buffer, "\\r", 2); break;
-                case '\t': fbuffer_append(out_buffer, "\\t", 2); break;
-                default:
-                    if (ch <= 0xFFFF) {
+                case '"':  FLUSH_POS(1); fbuffer_append(out_buffer, "\\\"", 2); break;
+                case '\\': FLUSH_POS(1); fbuffer_append(out_buffer, "\\\\", 2); break;
+                case '/':  FLUSH_POS(1); fbuffer_append(out_buffer, "\\/", 2); break;
+                case '\b': FLUSH_POS(1); fbuffer_append(out_buffer, "\\b", 2); break;
+                case '\f': FLUSH_POS(1); fbuffer_append(out_buffer, "\\f", 2); break;
+                case '\n': FLUSH_POS(1); fbuffer_append(out_buffer, "\\n", 2); break;
+                case '\r': FLUSH_POS(1); fbuffer_append(out_buffer, "\\r", 2); break;
+                case '\t': FLUSH_POS(1); fbuffer_append(out_buffer, "\\t", 2); break;
+                default: {
+                    if ((ch & 0x80) == 0x00) { /* leading 1 bit is   0b0     */
+                        FLUSH_POS(1);
                         scratch[2] = hexdig[ch >> 12];
                         scratch[3] = hexdig[(ch >> 8) & 0xf];
                         scratch[4] = hexdig[(ch >> 4) & 0xf];
                         scratch[5] = hexdig[ch & 0xf];
                         fbuffer_append(out_buffer, scratch, 6);
+                    } else if ((ch & 0xE0) == 0xC0) { /* leading 3 bits are 0b110   */
```

Comment:
Makes me wonder if we could actually store the character length in the escape table, and directly jump. 🤔

Reply:
31d811faf1baafb75c7bb5de0b20bae28a277c63

Reply:
One thing to care is this doesn't guarantee there are 1-6 bytes left in the String if the String is broken.

Reply:
We never enter this path if the string is broken: https://github.com/ruby/json/blob/9ce215d5e1e4a07d67036238a33924323fa78383/ext/json/ext/generator/generator.c#L739

We rely on the string coderange to guarantee us we don't have to deal with broken UTF-8.

The assumption is that most strings have their coderange known. If we wanted to avoid scanning the coderange when it is unknown we could have an alternative function that does protect against broken strings, but not sure if it's worth it.

Reply:
Ah right a broken String is just an exception on JSON.dump (unlike on JSON.load), nice.